### ISSUE TITLE
Generate signed URLs when using workload identity in Java

### ISSUE THREAD

**Issue by pedrosdf** [googleapis/google-cloud-java#10464]
Thanks for stopping by to let us know something could be better!

*PLEASE READ*: If you have a support contract with Google, please create an issue in the [support console](https://cloud.google.com/support/) instead of filing on GitHub. This will ensure a timely response.

*Is your feature request related to a problem? Please describe.*
[Storage] We are opening this Feature Request as we want to be able to easily generate signed URLs when using workload identity. We understand that this is a known missing feature and it is being worked on through other GitHub feature requests for .net [1] and ruby [2], but we want this functionality to also be available for java. 

*Describe the solution you'd like*
We found a recent Google Cloud Collective response in a Stack Overflow issue [3] explaining the following: “External account credentials (Workload ID) are not supported as URL signers and you need to use the IAM service to sign the blob yourself. External account credentials are not currently supported for URL signing because it's not always possible to know client side which service account the credential maps back to, and that's a requirement (we would be calling the IAM service internally for this).”
We know this is already being worked on and there are some workarounds for .net [1] and ruby [2], therefore, we want to have a resolution/workaround on the Java side as well.  

*Describe alternatives you've considered*
Tried using this example [4] by adapting it to Java, but started receiving the following exception:
Exception in thread "main" java.lang.IllegalStateException: Signing key was not provided and could not be derived
at com.google.common.base.Preconditions.checkState(Preconditions.java:512)
at com.google.cloud.storage.StorageImpl.signUrl(StorageImpl.java:709)
at com.ayla.cloud.gcs.controller.GCSController.getDownloadURL(GCSController.java:70)
at com.ayla.cloud.gcs.controller.GCSController.greeting(GCSController.java:42)
at com.ayla.cloud.gcs.controller.GCSController.main(GCSController.java:49)

This is the code being used:
package com.ayla.cloud.gcs.controller;
import com.google.auth.oauth2.ComputeEngineCredentials;
import com.google.auth.oauth2.IdTokenCredentials;
import com.google.cloud.storage.BlobId;
import com.google.cloud.storage.BlobInfo;
import com.google.cloud.storage.Storage;
import com.google.cloud.storage.StorageOptions;
import java.io.IOException;
import java.net.URL;
import java.util.Arrays;
import java.util.concurrent.TimeUnit;

public class GCSController {
public String greeting() {
String bucketName = "";
String path = "example2/example-1001";
long ttl = 60;
String gcsSignatureVersion = "V4";
ComputeEngineCredentials caCreds = ComputeEngineCredentials.create();
String targetAudience = "https://example.com";
IdTokenCredentials tokenCredential = IdTokenCredentials.newBuilder() .setIdTokenProvider(caCreds)
.setTargetAudience(targetAudience)
.setOptions(Arrays.asList(ComputeEngineCredentials.Option.FORMAT_FULL)) .build();
URL url = getDownloadURL(bucketName, path, ttl, gcsSignatureVersion, tokenCredential);
return url.toString();
}
public static void main(String[] args) throws IOException {
System.out.println(new GCSController().greeting());
}
public URL getDownloadURL(String bucketName, String path, long ttl, final String gcsSignatureVersion,

IdTokenCredentials tokenCredential) {
BlobInfo blobInfo = BlobInfo.newBuilder(BlobId.of(bucketName, path)).build();
Storage.SignUrlOption signUrlOption = null;

switch (gcsSignatureVersion) {
case "V2":
signUrlOption = Storage.SignUrlOption.withV2Signature();
break;
case "V4":
default:
signUrlOption = Storage.SignUrlOption.withV4Signature();
}
if (null == signUrlOption) {
return null;
}
Storage storage = StorageOptions.newBuilder().setCredentials(tokenCredential).build().getService();
return storage.signUrl(blobInfo, ttl, TimeUnit.MINUTES, signUrlOption);
}
}

*Additional context*
[1] https://github.com/googleapis/google-api-dotnet-client/issues/2410 
[2] https://github.com/googleapis/google-cloud-ruby/issues/13307 
[3] https://stackoverflow.com/a/76266912
[4] https://gist.github.com/jezhumble/91051485db4462add82045ef9ac2a0ec

**Comment by zhumin8** [comment]
Feature request to java-storage. 
@BenWhitehead Do you mind taking a look?

**Comment by BenWhitehead** [comment]
The code in google-cloud-storage to sign urls depends upon the value provided in `com.google.cloud.storage.Storage.SignUrlOption#signWith`, if that is not provided then it will attempt to fallback and use the instance of credentials provided at client creation time. The library specifically invokes [`com.google.auth.ServiceAccountSigner#sign(byte[])`](https://github.com/googleapis/google-auth-library-java/blob/23c3cbe70fdce49a3075e15ba965739704a87ace/credentials/java/com/google/auth/ServiceAccountSigner.java#L83) [here](https://github.com/googleapis/java-storage/blob/838f9f86edb674aa3352ba2f6c6011bc2f5c5ea7/google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java#L798).

In this case it sounds like the type of credentials provided doesn't know how to sign.

I think this is actually an issue that would need to be addressed in google-auth-library, rather than the storage library code itself.

@timursadykov Do you know if there is anything already on the roadmap to implement the sign method for credentials that work with workload identity?

**Comment by TimurSadykov** [comment]
ack, I'll chat with the team that owns WIF

**Comment by kolstae** [comment]
Any news on this?

**Comment by ViliusS** [comment]
Apparently there is a confusion in a lot of places what works and what isn't.

Standard Google Cloud Java libraries rely on `https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/{serviceAccountEmail}:signBlob` endpoint to generate the signature for URL. This endpoint only accepts service account email (or an ID) of the service account which exists in the Google infrastructure. Passing any other email address or a string (like `principal://...` of the Workload Identity ID) will result in 404 error.

Unless this is implemented at Google infrastructure level, external identities for URL signing won't be supported any time soon in any of the Google libraries.

However, this doesn't mean workload identities cannot be used at all. For example, Workload Identity Federation for GKE can be configured to use [service account impersonation](https://docs.cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#kubernetes-sa-to-iam) (also called account linking). When calling signUrl() from Google Cloud Java library app running on GKE it uses ADC and generates ComputeCredentials Auth object by querying GKE metadata server for the service account email. This email is then used to generate signature and sign URL.

The same goes for Workload Identity Federation with other OIDC providers. For example on GitHub you can use [the same service account impersonation mechanics](https://github.com/google-github-actions/auth#indirect-wif) to generate signed URLs.

This is not so secure as direct federated identity, but still no static service key credentials are needed.

So to summarize, the following configurations currently **works** to generate signed URLs without any changes:
* Workload Identity Federation for GKE with account linking (tested)
* Workload Identity Federation with any other OIDC provider which supports Service Account impersonation (untested)

The following configurations **do not work**:
* Workload Identity Federation for GKE with direct principals (tested)
* Workload Identity Federation with any other OIDC provider which generates federated tokens and directly authenticates to Google Cloud APIs (untested)
